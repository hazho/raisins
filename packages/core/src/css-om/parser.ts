import csstree from 'css-tree';

/**
 * Parses a CSS expression and returns a pure js object AST representation of the content
 *
 * @returns
 */
export function parser(css: string): csstree.CssNodePlain {
  var obj = csstree.parse(css, {});
  const plain = csstree.toPlainObject(obj);
  // Without this stringify->parse loop, there ends up being mutable references
  const extraPlain = JSON.parse(JSON.stringify(plain));
  return extraPlain;
}

export default parser;