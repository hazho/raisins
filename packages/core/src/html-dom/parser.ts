import htmlparser2 from 'htmlparser2';
import { domHandlerToRaisin } from './DomHandlerToRaisin';
import type { RaisinDocumentNode } from './RaisinNode';
import { removeWhitespace } from './util';

/**
 * Parses HTML into a RaisinDocumentNode.
 * 
 * Internally uses `htmlparser2` and then converts the `dom-handler`
 * document returned into a frozen immutable RaisinDocumentNode.
 * 
 * Also cleans 
 * 
 * @param html an HTML string
 * @returns a parsed RaisinDocumentNode
 */
export function parse(html: string): RaisinDocumentNode {
  const DomNode = htmlparser2.parseDocument(html);
  const raisinNode = domHandlerToRaisin(DomNode) as RaisinDocumentNode;

  const clean = removeWhitespace(raisinNode);
  return clean as RaisinDocumentNode;
}

export default parse;