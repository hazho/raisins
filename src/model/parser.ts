import { domHandlerToRaisin } from './DomHandlerToRaisin';
import { RaisinDocumentNode } from './RaisinNode';
import htmlparser2 from 'htmlparser2';
import { removeWhitespace } from '../util';

export function parse(html: string): RaisinDocumentNode {
  const DomNode = htmlparser2.parseDocument(html);
  const raisinNode = domHandlerToRaisin(DomNode) as RaisinDocumentNode;

  const clean = removeWhitespace(raisinNode);
  return clean as RaisinDocumentNode;
}
