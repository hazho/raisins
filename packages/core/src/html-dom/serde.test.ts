import { parse } from "./parser";
import serializer from "./serializer";

describe("Parse + serialize", () => {
  test("Can parse simple HTML", () => {
    const source = `<div>Example</div>`;
    const node = parse(source);
    expect(node).toMatchSnapshot();

    const out = serializer(node);

    expect(out).toBe(source);
  });
});
