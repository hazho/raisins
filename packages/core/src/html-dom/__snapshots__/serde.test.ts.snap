// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Parse + serialize Can parse simple HTML 1`] = `
Object {
  "children": Array [
    Object {
      "attribs": Object {},
      "children": Array [
        Object {
          "data": "Example",
          "nodeType": 3,
          "type": "text",
        },
      ],
      "nodeType": 1,
      "style": undefined,
      "tagName": "div",
      "type": "tag",
    },
  ],
  "nodeType": 9,
  "type": "root",
}
`;
